core/templates/dev/head/pages/exploration_editor/editor_tab/StateHints.js

```diff
@@ -0,0 +1,158 @@
+    alertsService, INTERACTION_SPECS, stateHintsService,
+    explorationStatesService, stateInteractionIdService,
+    UrlInterpolationService, HintObjectFactory) {
+    $scope.editorContextService = editorContextService;
+    $scope.stateHintsService = stateHintsService;
+    $scope.activeHintIndex = null;
+
+    $scope.dragDotsImgUrl = UrlInterpolationService.getStaticImageUrl(
+      '/general/drag_dots.png');
+
+    $scope.$on('stateEditorInitialized', function(evt, stateData) {
+      stateHintsService.init(
+        editorContextService.getActiveStateName(),
+          stateData.interaction.hints);
+
+      $scope.activeHintIndex = null;
+    });
+
+    $scope.getHintSummary = function(hint) {
+      var hintAsPlainText = (
+        hint.hint_text.length ?
+          $filter('convertToPlainText')(hint.hint_text) : '');
+      return hintAsPlainText;
+    };
+
+    $scope.changeActiveHintIndex = function(newIndex) {
+    // If the current hint is being clicked on again, close it.
+      if (newIndex === $scope.activeHintIndex) {
+        $scope.activeHintIndex = null;
+      } else {
+        $scope.activeHintIndex = newIndex;
+      }
+    };
+
+    // This returns false if the current interaction ID is null.
+    $scope.isCurrentInteractionLinear = function() {
+      var interactionId = stateInteractionIdService.savedMemento;
+      return interactionId && INTERACTION_SPECS[interactionId].is_linear;
+    };
+
+    $scope.openAddHintModal = function() {
+      alertsService.clearWarnings();
+      $rootScope.$broadcast('externalSave');
+
+      $modal.open({
+        templateUrl: 'modals/addHint',
+        backdrop: true,
+        controller: [
+          '$scope', '$modalInstance', 'editorContextService',
+          function($scope, $modalInstance, editorContextService) {
+            $scope.HINT_FORM_SCHEMA = {
+              type: 'html',
+              ui_config: {}
+            };
+
+            $scope.tmpHint = HintObjectFactory.createDefault('');
+
+            $scope.addHintForm = {};
+
+            $scope.hintIndex = stateHintsService.displayed.length + 1;
+
+            $scope.saveHint = function() {
+              // Close the modal and save it afterwards.
+              $modalInstance.close({
+                hint: angular.copy($scope.tmpHint)
+              });
+            };
+
+            $scope.cancel = function() {
+              $modalInstance.dismiss('cancel');
+              alertsService.clearWarnings();
+            };
+          }
+        ]
+      }).result.then(function(result) {
+        stateHintsService.displayed.push(result.hint);
+        stateHintsService.saveDisplayedValue();
+      });
+    };
+
+    // When the page is scrolled so that the top of the page is above the
+    // browser viewport, there are some bugs in the positioning of the helper.
+    // This is a bug in jQueryUI that has not been fixed yet. For more details,
+    // see http://stackoverflow.com/q/5791886
```

Comment:
I tried correcting the offset, but it looks like this offset keeps changing as new hints are added. So, I've let it be as-is for now. One thing I noted was zooming out a little in the browser fixes this offset issue.